### DISCUSSION TITLE
Message gets lost if sent immediately after the open event

### DISCUSSION THREAD

**Discussion by lorenzo-arena** [aiortc/aiortc#1233] [Q&A]
I have an application which consists of two parts, one of which is typically "the offerer", which also decides which datachannels to open, and the other waits for channels to be opened for communication. When a channel is opened on "the answerer" it has to do some operations (which are fast, but not immediate) before attaching a callback to the `message` event.

If I try to send a message right after receiving the `open` event in the offerer, like in the following snipped

```py
channel = pc.createDataChannel('my_custom_label')

@channel.on("message")
def on_message(message):
    logger.info(f"Received message with len {len(message)}, {message}")

@channel.on("open")
async def on_open():
    logger.info("Channel is open!")
    logger.info(f"Channel: sending message {my_message}")
    channel.send(my_message)
```

 the message seems to get lost. Is this expected behavior?

**Comment by MoreTore** [comment]
No, is not expected to lose the message on an open data channel. If thats all you wanted to know, im sure chatgpt would have told you the same thing.

**Comment by shahfaizanali** [comment]
@lorenzo-arena have you tried without using async?

**Comment by lorenzo-arena** [comment]
Sorry for the silence; we initially thought to have this solved by using a simple queue of messages received while setting up the channel; however it seems that the issue is related to the fact that if an `async` method is used as the `datachannel` event handler, its execution is effectively postponed an d if a message arrives on the channel in the meantime it is not managed by any handler.

I've opened a PR for this: https://github.com/aiortc/aiortc/pull/1394
